tests/e2e/etcd_grpcproxy_test.go

```diff
@@ -0,0 +1,163 @@
+// Copyright 2017 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package e2e
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+	"go.etcd.io/etcd/api/v3/etcdserverpb"
+	"go.etcd.io/etcd/pkg/v3/expect"
+	"go.etcd.io/etcd/tests/v3/framework/e2e"
+)
+
+func TestGrpcProxyAutoSync(t *testing.T) {
```

Comment:
I would suggest to add a test into [gateway_test.go](https://github.com/etcd-io/etcd/blob/main/tests/e2e/gateway_test.go) and  reuse some existing code, and also reuse [ctlV3MemberAdd](https://github.com/etcd-io/etcd/blob/1851316519eea48ef350fa642e8aaa5a8ba6b5d4/tests/e2e/ctl_v3_member_test.go#L193) or [ctlV3MemberRemove](https://github.com/etcd-io/etcd/blob/1851316519eea48ef350fa642e8aaa5a8ba6b5d4/tests/e2e/ctl_v3_member_test.go#L176)

I believe it can greatly simplify the implementation.

Reply:
Actually a better approach is to add a common test, which can be executed on both integration and e2e test environment. FYI. https://github.com/etcd-io/etcd/issues/13637

Reply:
1. I'm not sure why grpc-proxy tests should be in the same file with the gateway tests? They are different things, and therefore it's impossible to reuse startGateway function, for example.
2. I tried to use ctl* function, but they accept ctlCtx which I don't have and can't create properly. It useless without setting up epc field of type e2e.EtcdProcessCluster, which in turn I can't use because I need to add/remove member, which is impossible with e2e.EtcdProcessCluster.
3. I really love your idea to use common framework to write tests once and then run them for both integration and e2e env. But for this particular case I have two concerns:
a. Neither e2e framework nor common framework have API for adding/removing members properly.
b. GRPC-proxy is a standalone application, I'm not sure if we at can talk about integration tests at all.

Reply:
I also thought how to write integration test for grpc-proxy, but auto sync is part of client. What I want to check is the code inside etcdmain, which cannot be tested in integration tests way (without actually ports binding, running additional processes, etc.).

So in fact my changes can be tested only in e2e tests, I guess.

Reply:
> I'm not sure why grpc-proxy tests should be in the same file with the gateway tests? They are different things, and therefore it's impossible to reuse startGateway function, for example.

Yes, you are right. Sorry I misread the filename and content. They are indeed different things. 




> So in fact my changes can be tested only in e2e tests, I guess.

Agreed.



> I tried to use ctl* function, but they accept ctlCtx which I don't have and can't create properly. It useless without setting up epc field of type e2e.EtcdProcessCluster, which in turn I can't use because I need to add/remove member, which is impossible with e2e.EtcdProcessCluster.

There are lots of examples in the existing e2e test cases. Please try to reuse the existing utilities/methods/functions, such as [testCtl](https://github.com/etcd-io/etcd/blob/1851316519eea48ef350fa642e8aaa5a8ba6b5d4/tests/e2e/ctl_v3_test.go#L214), [getMemberList](https://github.com/etcd-io/etcd/blob/1851316519eea48ef350fa642e8aaa5a8ba6b5d4/tests/e2e/ctl_v3_member_test.go#L94), [ctlV3MemberAdd](https://github.com/etcd-io/etcd/blob/1851316519eea48ef350fa642e8aaa5a8ba6b5d4/tests/e2e/ctl_v3_member_test.go#L193), [ctlV3MemberRemove](https://github.com/etcd-io/etcd/blob/1851316519eea48ef350fa642e8aaa5a8ba6b5d4/tests/e2e/ctl_v3_member_test.go#L176), etc.

Reply:
Used e2e.EtcdctlV3, now it looks better (without custom json parser).

As for ctlV3 - I'm afraid I can't use it. Let me explain myself:
- gprc-proxy is not part of ctl utils, it's part of main etcd.
- All ctl* test function requires ctlCtx, which could be created only with e2e.EtcdProcessCluster.
- I can't use e2e.EtcdProcessCluster (neither with ctl*, nor directly), because it spawns all etcd nodes by its own. It's possible to change member list, but impossible to actually start new node or stop another.
- I also can't use e2e.EtcdProcess, because it could be created only from e2e.EtcdServerProcessConfig, which has private field `lg`, which is set only by function that creates e2e.EtcdProcessCluster, which I can't use (previous paragraph).
- At last, I found [number of examples](https://github.com/etcd-io/etcd/search?q=e2e.BinDir+etcd) in e2e tests, where nodes are spawned explicitly. So I hope for now my test fits into other e2e tests (:

Reply:
> * It's possible to change member list, but impossible to actually start new node or stop another

It seems like a real gap. Please consider to improve the existing e2e test framework if you have bandwidth and interested.

Reply:
BTW I found another place where adding new member would be useful: https://github.com/etcd-io/etcd/blob/main/tests/e2e/ctl_v3_snapshot_test.go#L258

Sure, I'll try to extend API.